From 86d05b4dde9492953e9564c09a3e3e74d1e9ce87 Mon Sep 17 00:00:00 2001
From: Justin Maggard <jmaggard10@gmail.com>
Date: Fri, 30 Jan 2015 12:09:20 -0800
Subject: [PATCH 06/19] Use friendly hostname for Avahi

 By default, Samba uses the NETBIOS name to advertise the system with Avahi.
 This makes it always appear uppercase, which is a little ugly.  Plus, we
 want to keep the protocol suffix.  So make those changes.
 .
 samba4 (2:4.0.5-netgear3) unstable; urgency=low
 .
   * Enable native Avahi support, but change default hostname displayed.
Author: Justin Maggard <justin.maggard@netgear.com>

---
 source3/smbd/avahi_register.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/source3/smbd/avahi_register.c b/source3/smbd/avahi_register.c
index 368168d..afc7adc 100644
--- a/source3/smbd/avahi_register.c
+++ b/source3/smbd/avahi_register.c
@@ -73,6 +73,18 @@ static void avahi_client_callback(AvahiClient *c, AvahiClientState status,
 	struct avahi_state_struct *state = talloc_get_type_abort(
 		userdata, struct avahi_state_struct);
 	int error;
+	char hn[64];
+
+	if (gethostname(hn, 56) == 0) {
+		char *c;
+		c = strchr(hn, '.');
+		if (c)
+			strncpy(c, " (SMB)", 8);
+		else
+			strncat(hn, " (SMB)", 8);
+	}
+	else
+		strncpy(hn, lp_netbios_name(), 32);
 
 	switch (status) {
 	case AVAHI_CLIENT_S_RUNNING:
@@ -88,7 +100,7 @@ static void avahi_client_callback(AvahiClient *c, AvahiClientState status,
 		}
 		if (avahi_entry_group_add_service(
 			    state->entry_group, AVAHI_IF_UNSPEC,
-			    AVAHI_PROTO_UNSPEC, 0, lp_netbios_name(),
+			    AVAHI_PROTO_UNSPEC, 0, hn,
 			    "_smb._tcp", NULL, NULL, state->port, NULL) < 0) {
 			error = avahi_client_errno(c);
 			DEBUG(10, ("avahi_entry_group_add_service failed: "
-- 
1.9.1

